### PR TITLE
feat(weave): OpenAI realtime support

### PR DESCRIPTION
**Pull request by zbirenbaum**
## Description

- Fixes WB-26886

Adds SDK support for OpenAI realtime API

Includes a number of autopatchers for common websockets libraries as well as the ability to instrument your own websocket and disable the autopatcher.

Couple of notes before this goes into review:
1. Types - Any is used a lot, I have a good set of types, they need to be integrated before this goes live
2. Testing - Testing this in CI is a bit difficult. I have a large number of integration tests but they don't always accurately reflect live behavior
3. Edge cases - Event order is implementation specific, the best solution I could come up with to get transcripts as much as possible was to add a debounce. There are a couple of other scenarios like this which may come up
4. Sessions.py is too long, probably want to split this up somehow

### CODE REVIEW

File: weave/integrations/openai_realtime/__init__.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# OpenAI Realtime API integration for Weave
+# This module automatically patches websocket connections to track sessions and conversations
+
+from .openai_realtime_sdk import (
```
Comment (neutralino1) [inline weave/integrations/openai_realtime/__init__.py]: Full import path please?

**Review by neutralino1 — COMMENTED** [review]
Amazing work.

You mention you have tests but I don't see them.

File: weave/trace/autopatch.py [hunk 2]
```diff
@@ -89,6 +90,9 @@ def autopatch(settings: Optional[AutopatchSettings] = None) -> None:
     from weave.integrations.notdiamond.tracing import get_notdiamond_patcher
     from weave.integrations.openai.openai_sdk import get_openai_patcher
     from weave.integrations.openai_agents.openai_agents import get_openai_agents_patcher
+    from weave.integrations.openai_realtime.openai_realtime_sdk import (
+        get_websocket_patcher,
```
Comment (andrewtruong) [inline weave/trace/autopatch.py]: this will change soon, but for now I recommend calling this specifically the `openai_realtime_websocket_patcher`

File: weave/integrations/openai_realtime/audio_buffer.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    sample_rate_hz: int = 24000
+    bits_per_sample: int = 16
+    channels: int = 1
+    buffer: bytearray = field(default_factory=bytearray)
+
+    def bytes_per_sample(self) -> int:
+        return (self.bits_per_sample // 8) * self.channels
+
+    def append_base64(self, b64: str) -> None:
```
Comment (andrewtruong) [inline weave/integrations/openai_realtime/audio_buffer.py]: shouldn't this be `extend`?

File: weave/trace/autopatch.py (hunk 2 above)
Comment (zbirenbaum) [reply weave/trace/autopatch.py]: Just to clarify:

~You mean to change `get_websocket_patcher` -> `openai_realtime_websocket_patcher` correct?~
Changed to 
```
from weave.integrations.openai_realtime import get_openai_realtime_websocket_patcher
```

File: weave/integrations/openai_realtime/audio_buffer.py (hunk 3 above)
Comment (zbirenbaum) [reply weave/integrations/openai_realtime/audio_buffer.py]: Good call, updated

File: weave/integrations/openai_realtime/models.py [hunk 4]
```diff
@@ -0,0 +1,931 @@
...
+SessionID = Annotated[str, AfterValidator(is_sess_id)]
+EventID = Annotated[str, AfterValidator(is_event_id)]
+ItemID = Annotated[str, AfterValidator(is_item_id)]
+ResponseID = Annotated[str, AfterValidator(is_resp_id)]
+CallID = Annotated[str, AfterValidator(is_call_id)]
+ConversationID = Annotated[str, AfterValidator(is_conv_id)]
+
+
+Voice = Literal["alloy", "shimmer", "echo"]
+AudioFormat = Literal["pcm16", "g711-ulaw", "g711-alaw"]
```
Comment (andrewtruong) [inline weave/integrations/openai_realtime/models.py]: please add docs for where these values came from!

File: weave/integrations/openai_realtime/models.py
Comment (andrewtruong) [inline weave/integrations/openai_realtime/models.py:1]: Are many of these types coming from OpenAI?  Why not use their actual types?

File: weave/integrations/openai_realtime/models.py [hunk 5]
```diff
@@ -0,0 +1,931 @@
...
+    "response.audio.done": ResponseAudioDoneMessage,
+    "response.function_call_arguments.delta": ResponseFunctionCallArgumentsDeltaMessage,
+    "response.function_call_arguments.done": ResponseFunctionCallArgumentsDoneMessage,
+    "rate_limits.updated": RateLimitsUpdatedMessage,
+}
+
+MessageType = Union[UserMessageType, ServerMessageType]
+
+
+def create_user_message_from_dict(data: dict) -> UserMessageType:
```
Comment (andrewtruong) [inline weave/integrations/openai_realtime/models.py]: wouldn't it be easier to use a discriminated union for all of these classes?
https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions-with-str-discriminators

File: weave/integrations/openai_realtime/models.py [hunk 6]
```diff
@@ -0,0 +1,931 @@
...
+    return UnknownClientMessage(**data)
+
+
+# Whenever we get a response we have to acquire a lock and associate it with the last conversation.item.recated item
+# Whenever we get a delta event we have to acquire a lock
+
+
+def has_item_id(message: MessageType) -> bool:
+    """Check if a message has an item ID, directly or nested on item.id."""
+    if hasattr(message, "item_id") and message.item_id is not None:
```
Comment (andrewtruong) [inline weave/integrations/openai_realtime/models.py]: nit: This can be a `if getattr(message, "item_id", None) is not None`

File: weave/integrations/openai_realtime/openai_realtime_websocket_patcher.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+import logging
+from functools import wraps
+from typing import TYPE_CHECKING, Any, Callable
+
+from weave.integrations.openai_realtime import connection
+from weave.integrations.patcher import MultiPatcher, NoOpPatcher, SymbolPatcher
+from weave.trace.autopatch import IntegrationSettings
+
+if TYPE_CHECKING:
+    pass
```
Comment (andrewtruong) [inline weave/integrations/openai_realtime/openai_realtime_websocket_patcher.py]: remove

File: weave/integrations/openai_realtime/models.py (hunk 6 above)
Comment (zbirenbaum) [reply weave/integrations/openai_realtime/models.py]: These functions were removed since they aren't really necessary

File: weave/integrations/openai_realtime/models.py
Comment (zbirenbaum) [reply weave/integrations/openai_realtime/models.py:1]: I attempted this, it's definitely not possible - Realtime is in super active development and OpenAI changes the types slightly on a weekly basis.

~~Essentially if we use official types we are praying that all of these are 100% up to date:~~
~~1. The OpenAI Realtime Server Endpoint~~
~~2. The user's client messages~~
~~3. The version of the SDK we got those types from~~

~~The user has no obligation to use the SDK with websockets and many clients will be written without it. Even if they have the package installed their messages likely will not 100% match the types in the SDK~~

Edit: OpenAI finally got around to updating their docs (they didn't when they took realtime out of beta) and it turns out the reason I couldn't get their types to work is because of a header flag that was previously required to function and now is not:

With Flag: You get messages matching beta API
Without Flag: You get messages matching the release API

I'll have to change this but for now lets get this in since it works for the beta release.

File: weave/integrations/openai_realtime/models.py (hunk 5 above)
Comment (zbirenbaum) [reply weave/integrations/openai_realtime/models.py]: This would be a followup. I'm going to hold off on non-functional changes until this goes in.

File: weave/integrations/openai_realtime/__init__.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+# This module automatically patches websocket connections to track sessions and conversations
+from weave.integrations.openai_realtime.openai_realtime_websocket_patcher import (
+    OpenAIRealtimeSettings,
+    get_openai_realtime_websocket_patcher,
+    wrap_websocket_async,
+    wrap_websocket_connect,
+    wrap_websocket_sync,
+)
+
+__all__ = [
```
Comment (tssweeney) [inline weave/integrations/openai_realtime/__init__.py:11]: are all of these truly "external" APIs? for example `wrap_*`?

File: weave/integrations/openai_realtime/audio_buffer.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    buffer: bytearray = field(default_factory=bytearray)
+
+    def bytes_per_sample(self) -> int:
+        return (self.bits_per_sample // 8) * self.channels
+
+    def extend_base64(self, b64: str) -> None:
+        """Decode base64 audio and extend buffer. Ignores invalid payloads gracefully."""
+        try:
+            self.buffer.extend(base64.b64decode(b64))
+        except Exception as e:
```
Comment (tssweeney) [inline weave/integrations/openai_realtime/audio_buffer.py]: bare exceptions are risky. If you truly don't know the exception you are looking for, please just wrap the decode in the try-catch to limit the over-exposure

File: weave/integrations/openai_realtime/models.py [hunk 10]
```diff
@@ -0,0 +1,902 @@
+from __future__ import annotations
```
Comment (tssweeney) [inline weave/integrations/openai_realtime/models.py:1]: i would add a comment to the header of this file describing why you had to do this and probably mentioning that it should not change / stay in sync with openai. i'd also recommend not calling it `models` due to the overloaded nature of that word in our app

File: weave/integrations/openai_realtime/models.py (hunk 4 above)
Comment (zbirenbaum) [reply weave/integrations/openai_realtime/models.py]: Added docs citation and updated voice list

File: weave/integrations/openai_realtime/__init__.py (hunk 8 above)
Comment (zbirenbaum) [reply weave/integrations/openai_realtime/__init__.py:11]: Yes, those are the methods for manual patching
```
        import websockets
        from weave.integrations.openai_realtime import wrap_websocket_async

        async def connect():
            ws = await websockets.connect(url)
            tracked_ws = wrap_websocket_async(ws)
            return tracked_ws
```

File: weave/integrations/openai_realtime/audio_buffer.py (hunk 9 above)
Comment (zbirenbaum) [reply weave/integrations/openai_realtime/audio_buffer.py]: addressed

File: weave/integrations/openai_realtime/models.py (hunk 10 above)
Comment (zbirenbaum) [reply weave/integrations/openai_realtime/models.py:1]: It's likely to change/be removed at some point. Especially once OpenAI manages to stabilize their API types and fix some stuff. I'm going to leave it as is for now but I'll add a comment
